metals/src/main/scala/scala/meta/internal/metals/JavaInteractiveSemanticdb.scala

```diff
@@ -0,0 +1,227 @@
+
+    val doc = if (semanticdbFile.exists) {
+      FileIO
+        .readAllDocuments(semanticdbFile)
+        .headOption
+        .getOrElse(s.TextDocument())
+    } else {
+      val log = stdout.result()
+      if (exitCode != 0 || log.nonEmpty)
+        scribe.warn(s"Javac-semanticdb output:\n${log.mkString("\n")}")
+      s.TextDocument()
+    }
+
+    val out = doc.copy(
+      uri = source.toURI.toString(),
+      text = text,
+      md5 = MD5.compute(text)
+    )
+
+    workDir.deleteRecursively()
+    out
+  }
+
+  private def patchModuleFlags(
+      source: AbsolutePath,
+      sourceRoot: AbsolutePath
+  ): List[String] = {
+    // Jigsaw doesn't allow compiling source with package
+    // that is declared in some existing module.
+    // It fails with: `error: package exists in another module: $packageName`
+    // but it might be fixed by passing `--patch-module $moduleName=$sourceRoot` option.
+    //
+    // Currently there is no infrastucture to detect if package belong to jigsaw module or not
+    // so this case is covered only for JDK sources.
+    if (jdkVersion.hasJigsaw) {
+      source.toRelativeInside(readonly) match {
+        case Some(rel) =>
+          val names = rel.toNIO.iterator().asScala.toList.map(_.filename)
+          names match {
+            case Directories.dependenciesName :: JdkSources.zipFileName :: moduleName :: _ =>
+              List("--patch-module", s"$moduleName=$sourceRoot")
+            case _ =>
+              Nil
+          }
+        case None => Nil
+      }
+    } else Nil
+  }
+
+  private def addExportsFlags: List[String] = {
+    if (jdkVersion.major >= 17) {
+      val compilerPackages = List(
+        "com.sun.tools.javac.api",
+        "com.sun.tools.javac.code",
+        "com.sun.tools.javac.model",
+        "com.sun.tools.javac.tree",
+        "com.sun.tools.javac.util"
+      )
+      compilerPackages.flatMap(pkg =>
+        List("-J--add-exports", s"-Jjdk.compiler/$pkg=ALL_UNNAMED")
+      )
+    } else Nil
+  }
+
+}
+
+object JavaInteractiveSemanticdb {
+
+  def create(
+      javaHome: AbsolutePath,
+      workspace: AbsolutePath,
+      buildTargets: BuildTargets
+  ): Option[JavaInteractiveSemanticdb] = {
+    val javac = javaHome.resolve("bin").resolve("javac")
+    getJavaVersionFromJavaHome(javaHome) match {
+      case Some(version) if javac.exists =>
+        val pluginJars = Embedded.downloadSemanticdbJavac
+        val instance = new JavaInteractiveSemanticdb(
+          javac,
+          version,
+          pluginJars,
+          workspace,
+          buildTargets
+        )
+        Some(instance)
+      case value =>
+        scribe.warn(
+          s"Can't instantiate JavaInteractiveSemanticdb (version: ${value}, javac exists: ${javac.exists})"
+        )
+        None
+    }
+  }
+
+  private def getJavaVersionFromJavaHome(
+      javaHome: AbsolutePath
+  ): Option[JdkVersion] = {
+
+    def fromReleaseFile: Option[JdkVersion] = {
```

Comment:
Wouldn't this be the same version as the currently running one?

Reply:
Hmm.. I'm bit lost in java homes.
Can `userConfig.javaHome` be different from process one?

One additional thing I've noticed is that BSP might have different jdk settings (`JvmBuildTarget.javaHome`) but we currently don't handle it.

Reply:
> Can userConfig.javaHome be different from process one?

`userConfig.javaHome` can be different if the user changed the setting while running, otherwise it should be the same.

> One additional thing I've noticed is that BSP might have different jdk settings (JvmBuildTarget.javaHome) but we currently don't handle it.

Yeah, we could get the java home from BSP I think and maybe use that instead for javaHome in case of definitions?

Reply:
> Yeah, we could get the java home from BSP I think and maybe use that instead for javaHome in case of definitions?

I would turn it into a seperate PR. This should work well for existing JdkSources mechanic that is taken from `userConfig.javaHome || process java home`. It's needed to switch them on jdk from BSP first. Otherwise there will be jdk version mismatch between javac in this class and the actual source that we pass to it.

Reply:
@tgodzik 
reported it here - #3479

Could you we merge as it is?
I think it will be ok to fix in the next release.